_posts/2019-06-25-building-a-mono-repo.markdown

```diff
@@ -0,0 +1,66 @@
+
+I will however run through three approaches I've used in the past and explain the pros and cons of each.
+
+#### 1. Bash Your Way Through It
+
+Start with a simple bash script that understands project dependencies, looks at the diff of the change set in question, and infers what needs to be built in what order. It grows and grows, and six months later you realize you've written your own build system. Weeping ensues.
+
+**_Pros_**: Your build is defined in your repo and so in theory should be owned by the whole team. Anyone can add a new project and configure its dependencies (hopefully).
+
+**_Cons_**: Loss of build transparency. Questions like "*how often does project Y get built*" or "*how long does project X take to build on average*" are difficult to answer. They likely require parsing build logs and trying to regex out the time to build project Y. Assuming you thought to print that info to begin with.
+
+You are frequently having to remind that one person who answers these questions with data from your Jenkins monojob that "*we build something different on each run*" so "*those numbers are basically meaningless*", and "Seriously Garry, we just talked about this last week*".
+
+#### 2. Directed A-Sickening Graph
+Give each of your projects their own Jekins job. Use the path regex option to restrict builds from triggering unless said project has been modified. Create dependencies between each of these jobs that models the Directed Acyclic Graph of your project dependencies. Setup the [dependency graph plugin](https://wiki.jenkins.io/display/JENKINS/Dependency+Graph+View+Plugin) after a small change to a `README` file unexpectedly causes every project to be rebuilt and deployed. Then react in horror as it shows you the monsterous dependnecy graph you're created. At least with a shell script no can actually see the mess.
+
+
+**_Pro_**: Transparency! For better or worse you'll always know when, where, and why each build has been triggered.
+
+**_Cons_**: You've taken complexity that will exist no matter what and moved it from your repo into Jenkins configuration that no one really understands... except maybe you? Adding a new project results in a slew of Jiras for Jenkins setup that are... wait, why are these assigned me? I thought we agree'd people would self service? Wait, I'm the only one who knows how to set these up? Ug, fine.
+
+Also, I haven't used this style of configuration in some time, and if you want handy dandy Pull Request evaluation, it probably won't work anyway.  Unless you attach every build to your PR's and shortcircuit them to not do anyting if a change isn't required. Which, if you do, congratulations, you've just married this approach with the "build your own build system" approach from above.
+
+#### 3. Words Mean What I Say They Mean
+Redefine your idea of a build *is* so that you can run through everything on every change. Build every one of your projects for every Pull Request.  How do you get there?
```

Comment:
I think you need to be more specific here? I dont really understand the idea, but I think you are proposing that the definition is wrong, but you haven't defined a new definition. "Redefine your idea of what a build even is - now a build doesn't mean <original def>, in this brave new world it means <new def>. Here's how - build every..."

Reply:
coming back after reading more of the details - I feel like you should call out that speed is a thing. "To get there, you'll need to configure things differently, unless you love day long build cycles in your *continuous* integration system. Here's some tips"